File: arrays/merge-sorted-arrays.js
const array1 = [0, 3, 4, 31];
const array2 = [4, 6, 30];

function mergeSortedArrays(arr1, arr2) {
  // check if the inputs are arrays
  // loop over each array and add the items to a  new array
  // the array with the items is then sorted
  if (Array.isArray(arr1) !== true || Array.isArray(arr2) !== true) {
    return "Incorrect input, Please use an array object";
  }

  const mergedArray = [];
  for (let i = 0; i < arr1.length; i++) {
    mergedArray.push(arr1[i]);
  }

  for (let i = 0; i < arr2.length; i++) {
    mergedArray.push(arr2[i]);
  }

  return mergedArray.sort((a, b) => a - b);
}

// console.log(mergeSortedArrays(array1, array2));

function mergeSortedArrays2(array1, array2) {
  // Typical assumption is that array is sorted
  // check inputs to confirm they have elements, if one of then don't, return the other
  const mergeArrays = [];
  let array1Item = array1[0];
  let array2Item = array2[0];
  let i = 0;
  let j = 0;

  if (array1.length === 0) {
    return arr2.sort();
  }

  if (array2.length === 0) {
    return arr1.sort();
  }

  if (Array.isArray(array1) !== true || Array.isArray(array2) !== true) {
    return "Incorrect input, Please use an array object";
  }

  while (array1Item || array2Item) {
    if (!array2Item || array1Item < array2Item) {
      console.log(array1Item, array2Item);
      mergeArrays.push(array1Item);
      i++;
      array1Item = array1[i];
    } else {
      mergeArrays.push(array2Item);
      j++;
      array2Item = array2[j];
    }
  }
  return mergeArrays;
}
console.log(mergeSortedArrays2([0, 3, 4, 31], [4, 6, 30]));

File: Data-structures/arrays/python-implementation/merge-sorted-arrays.py
array_1 = [0, 3, 4];
array_2 = [0, 4, 6, 30];

# def merge_sorted_arrays(array_1, array_2):
#     merged_array = []
#     array_1_item = array_1[0]
#     array_2_item = array_2[0]
#     i = 0
#     j = 0 
#     while (array_1_item | array_2_item):
#         print(array_1_item, array_2_item)
#         if ((not array_2_item) | (array_1_item < array_2_item)):
#             merged_array.append(array_1_item)
#             i += 1
#             try:
#                 array_1_item = array_1[i]
#             except:
#                 array_1_item = False
#         else:
#             merged_array.append(array_2_item)
#             j += 1 
#             try:
#                 array_2_item = array_2[j]
#             except:
#                 array_2_item = False

#     return merged_array

# print(merge_sorted_arrays(array_1, array_2))
# print(False | False)


# The below function works provided that none of the inputs is 0
def merge_sorted_arrays_2(array_1, array_2):
    merged_array = []
    array_1_item = array_1[0]
    array_2_item = array_2[0]
    i = 0
    j = 0 

    if (max(array_1) < max(array_2)):
        while (array_1_item | array_2_item):
            print(array_1_item, array_2_item)

            if (((not (array_1_item)) | (array_1_item == 0)) | (array_2_item < array_1_item)):
                merged_array.append(array_2_item)
                j += 1
                try:
                    array_2_item = array_2[j]
                except:
                    array_2_item = False
            else:
                merged_array.append(array_1_item)
                i += 1 
                try:
                    array_1_item = array_1[i]
                except:
                    array_1_item = False
    else:
        while (array_1_item | array_2_item):
            print(array_1_item, array_2_item)
            if (((not (array_2_item)) | (array_2_item == 0)) | (array_1_item < array_2_item)):
                merged_array.append(array_1_item)
                i += 1
                try:
                    array_1_item = array_1[i]
                except:
                    array_1_item = False
            else:
                merged_array.append(array_2_item)
                j += 1 
                try:
                    array_2_item = array_2[j]
                except:
                    array_2_item = False

    return merged_array



print(merge_sorted_arrays_2(array_1, array_2))
# print(min(array_1))
# print(min(array_2))





